### PR TITLE
vLLM support for classifier reward models

### PR DESCRIPTION
**Pull request by rohingarg-c**
support for multiple named classifier heads, allow arbitrary data (besides just the hidden states) to be passed only to the classifier head, fixed some small transformer backend bugs, etc.

### CODE REVIEW

File: vllm/model_executor/models/transformers.py [hunk 1]
```diff
@@ -410,10 +411,12 @@ def forward(
     def compute_additional_head(
         self,
         hidden_states: torch.Tensor,
+        additional_heads_data: Optional[dict[str, Any]],
```
Comment (renou) [inline vllm/model_executor/models/transformers.py:414]: maybe call it additional_heads_extra_features? Or additional_heads_extra_inputs?

File: vllm/v1/worker/gpu_input_batch.py [hunk 2]
```diff
@@ -286,11 +287,11 @@ def add_request(
         num_prompt_tokens = len(request.prompt_token_ids)
         self.num_prompt_tokens[req_index] = num_prompt_tokens
         self.token_ids_cpu[
-        req_index, :num_prompt_tokens] = request.prompt_token_ids
+            req_index, :num_prompt_tokens] = request.prompt_token_ids
```
Comment (renou) [inline vllm/v1/worker/gpu_input_batch.py:290]: looks like unintended changes?

File: vllm/v1/worker/gpu_input_batch.py [hunk 3]
```diff
@@ -286,11 +287,11 @@ def add_request(
...
         start_idx = num_prompt_tokens
         end_idx = start_idx + len(request.output_token_ids)
         self.token_ids_cpu[req_index,
-        start_idx:end_idx] = request.output_token_ids
+                           start_idx:end_idx] = request.output_token_ids
```
Comment (renou) [inline vllm/v1/worker/gpu_input_batch.py:294]: looks like unintended changes?

File: vllm/v1/worker/gpu_model_runner.py [hunk 4]
```diff
@@ -1429,16 +1429,25 @@ def execute_model(
 
         if run_additional_heads:
             assert hasattr(self.model, "compute_additional_head")
+            additional_heads_data: Optional[list[Optional[dict[str, Any]]]] = [
```
Comment (renou) [inline vllm/v1/worker/gpu_model_runner.py:1432]: Do we need the second Optional? It allows `value = [None, {"loss": 0.12}, {"extra": [1, 2, 3]}, None]`.

File: vllm/entrypoints/openai/protocol.py [hunk 5]
```diff
@@ -1375,47 +1374,6 @@ class ScoreResponse(OpenAIBaseModel):
     usage: UsageInfo
 
 
-class ClassificationRequest(OpenAIBaseModel):
```
[the remark this answers is not in the document]
Comment (rohingarg-c) [reply vllm/entrypoints/openai/protocol.py:1378]: yes, the code is duplicated in the file and was giving mypy errors (i think it was an artifact of merging vllm main into our fork)

File: vllm/model_executor/model_loader/default_loader.py [hunk 6]
```diff
@@ -278,5 +278,7 @@ def load_weights(self, model: nn.Module,
         if model_config.quantization is None and loaded_weights is not None:
             weights_not_loaded = weights_to_load - loaded_weights
             if weights_not_loaded:
-                raise ValueError("Following weights were not initialized from "
-                                 f"checkpoint: {weights_not_loaded}")
+                # raise ValueError("Following weights were not initialized "
+                #                  f"from checkpoint: {weights_not_loaded}")
+                logger.error("Following weights were not initialized "
```
[the remark this answers is not in the document]
Comment (rohingarg-c) [reply vllm/model_executor/model_loader/default_loader.py:283]: if you throw an error instead of just logging it, it would crash when lm_head wasn't supplied (since transformers backend for vLLM expects it). however, it's not needed for classifier heads, so i converted the message to a logger error instead of it crashing. not sure if there's a better way of handling it.

File: vllm/model_executor/models/transformers.py [hunk 7]
```diff
@@ -306,11 +306,12 @@ def create_attention_instances(self) -> dict[int, Attention]:
                 self.config.global_attention_layers, list):
             global_attention_layers = self.config.global_attention_layers
         else:
-            global_attention_layers = []
+            global_attention_layers = None
```
[the remark this answers is not in the document]
Comment (rohingarg-c) [reply vllm/model_executor/models/transformers.py:309]: Small distinction; `None` implies that **all** layers are global attention layers while `[]` would imply that **none** of them are. It's relevant for the change later on line 313 that checks if the index is in the global_attention_layers array to determine if it's a sliding window attention layer or not.

File: vllm/model_executor/models/transformers.py (hunk 1 above)
Comment (rohingarg-c) [reply vllm/model_executor/models/transformers.py:414]: `additional_heads_extra_inputs` sounds good to me, will change.

File: vllm/v1/worker/gpu_model_runner.py (hunk 4 above)
Comment (rohingarg-c) [reply vllm/v1/worker/gpu_model_runner.py:1432]: this was intended behavior. it was written that way so that for models that don't require additional inputs `None` would be acceptable. I assume this will be handled on the model side. the first optional was so that, in cases where no additional input was being passed in, the model could process the whole batch at once rather than each request separately for a bit of efficiency.

File: vllm/model_executor/model_loader/default_loader.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rohingarg-c) [reply vllm/model_executor/model_loader/default_loader.py:283]: linear layer that maps hidden states to token ids

File: vllm/v1/worker/gpu_model_runner.py (hunk 4 above)
Comment (renou) [reply vllm/v1/worker/gpu_model_runner.py]: I see, so:
1. for models that don't require additional inputs: `additional_heads_data` would be `None` for entire batch.
2. for models that require additional inputs: we still allow in the same batch, some rows don't have extra data, some rows have extra data?

File: vllm/entrypoints/openai/protocol.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (renou) [reply vllm/entrypoints/openai/protocol.py:1378]: Thanks!

File: vllm/model_executor/model_loader/default_loader.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (renou) [reply vllm/model_executor/model_loader/default_loader.py:283]: Got it. Thanks! Although it's also ok to load it?

File: vllm/v1/worker/gpu_model_runner.py (hunk 4 above)
Comment (renou) [reply vllm/v1/worker/gpu_model_runner.py]: Make sense. Thanks!

**Review by renou — APPROVED** [review]

File: vllm/model_executor/models/transformers.py [hunk 8]
```diff
@@ -514,9 +521,11 @@ def compute_logits(
     def compute_additional_head(
         self,
         hidden_states: torch.Tensor,
+        additional_heads_extra_inputs: Optional[list[dict[str, Any]]],
```
Comment (renou) [inline vllm/model_executor/models/transformers.py:524]: Based on how I implemented in reward model (https://github.com/character-tech/character-tech/pull/24365/files), we may want to update the signature here.
image

Comment (rohingarg-c) [reply vllm/model_executor/models/transformers.py:524]: do you think it would make sense to do the `list[dict[str, Any]] -> dict[str, torch.Tensor]` conversion inside the reward model rather than inside vLLM? the extra input doesn't necessarily fit cleanly in a tensor right (e.g. potentially the data isn't all the same length)? Moving it earlier in the chain doesn't improve efficiency either.

Comment (renou) [reply vllm/model_executor/models/transformers.py:524]: discussed offline. We will convert in a separate wrapper class between vllm and reward model.
